### PR TITLE
config.Plugins should not be composed of pointers

### LINKED ISSUE #95190: config.Plugins should not be composed of pointers
There is no benefit of having `pkg/scheduler/apis/config.Plugins` be composed of Pointers rather than concrete `PluginSet`s

We require configurations to explicitly set `disabled: ["*"]` if they want to indicate that they don't want the default plugins. We even create an empty `PluginSet` when we are merging plugin configurations

https://github.com/kubernetes/kubernetes/blob/aa1f91286838d373a6d9217ad9f4fe0647a1c859/pkg/scheduler/apis/config/types.go#L307-L310

The inconvenience of having them as pointers is that sometimes we have to use `nil` and sometimes an empty `PluginSet` in unit tests.

/sig scheduling
/kind cleanup

### PR DESCRIPTION
**Pull request by arghya88**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #95190

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Comment by arghya88** [comment]
cc @alculquicondor

File: staging/src/k8s.io/kube-scheduler/config/v1beta1/types.go [hunk 1]
```diff
@@ -155,39 +155,39 @@ type KubeSchedulerProfile struct {
 // be invoked before default plugins, default plugins must be disabled and re-enabled here in desired order.
 type Plugins struct {
 	// QueueSort is a list of plugins that should be invoked when sorting pods in the scheduling queue.
-	QueueSort *PluginSet `json:"queueSort,omitempty"`
```
Comment (alculquicondor) [inline staging/src/k8s.io/kube-scheduler/config/v1beta1/types.go:158]: revert the changes in this file. It is actually convenient to have pointers here so that we can really "omitempty".
We might need custom converstion. We will see. Revert this file first.

File: pkg/scheduler/testing/framework_helpers.go [hunk 2]
```diff
@@ -131,9 +131,9 @@ func getPluginSetByExtension(plugins *schedulerapi.Plugins, extension string) *s
 	}
 }
 
-func initializeIfNeeded(s **schedulerapi.PluginSet) *schedulerapi.PluginSet {
-	if *s == nil {
-		*s = &schedulerapi.PluginSet{}
+func initializeIfNeeded(s *schedulerapi.PluginSet) *schedulerapi.PluginSet {
```
Comment (alculquicondor) [inline pkg/scheduler/testing/framework_helpers.go:134]: shouldn't this function go away?

File: pkg/scheduler/apis/config/types_test.go [hunk 3]
```diff
@@ -224,59 +224,50 @@ func TestPluginsApply(t *testing.T) {
...
 				},
-				PostFilter: &PluginSet{Enabled: []Plugin{}},
-				PreScore:   &PluginSet{Enabled: []Plugin{}},
-				Score:      &PluginSet{Enabled: []Plugin{}},
-				Reserve:    &PluginSet{Enabled: []Plugin{}},
-				Permit:     &PluginSet{Enabled: []Plugin{}},
-				PreBind:    &PluginSet{Enabled: []Plugin{}},
-				Bind:       &PluginSet{Enabled: []Plugin{}},
-				PostBind:   &PluginSet{Enabled: []Plugin{}},
+				PostFilter: PluginSet{Enabled: []Plugin{}},
```
Comment (alculquicondor) [inline pkg/scheduler/apis/config/types_test.go:243]: Actually, the objective should be to get rid of these lines.

They should be something like: `PostFilter PluginSet{Enabled: nil, Disabled: nil}`, which you can actually just not remove.

File: pkg/scheduler/apis/config/types.go [hunk 4]
```diff
@@ -331,7 +320,7 @@ func mergePluginSets(defaultPluginSet, customPluginSet *PluginSet) *PluginSet {
 
 	enabledPlugins = append(enabledPlugins, customPluginSet.Enabled...)
```
Comment (alculquicondor) [inline pkg/scheduler/apis/config/types.go:321]: in line 310, do:

`var enabledPlugins []Plugin`

This will keep it `nil` until an item is appended

File: pkg/scheduler/apis/config/validation/validation_pluginargs_test.go [hunk 5]
```diff
@@ -513,7 +514,7 @@ func assertErr(t *testing.T, wantErr string, gotErr error) {
 		if gotErr == nil {
 			t.Fatalf("\nwant err to be:\n\t%s\ngot:\n\tnil", wantErr)
 		}
-		if gotErr.Error() != wantErr {
+		if !errors.Is(gotErr, errors.New(wantErr)) {
```
Comment (alculquicondor) [inline pkg/scheduler/apis/config/validation/validation_pluginargs_test.go:517]: while I like this, it's unrelated. Could you do it in a separate PR, you can reference #95047

File: pkg/scheduler/testing/framework_helpers.go (hunk 2 above)
Comment (alculquicondor) [reply pkg/scheduler/testing/framework_helpers.go:134]: Any comments on this?

File: pkg/scheduler/apis/config/types_test.go (hunk 3 above)
Comment (arghya88) [reply pkg/scheduler/apis/config/types_test.go:243]: Wondering what to do here

File: pkg/scheduler/apis/config/types_test.go [hunk 6]
```diff
@@ -32,38 +32,38 @@ func TestPluginsAppend(t *testing.T) {
...
 				},
-				PostFilter: &PluginSet{},
-				PreScore:   &PluginSet{},
-				Score:      &PluginSet{},
-				Reserve:    &PluginSet{},
-				Permit:     &PluginSet{},
-				PreBind:    &PluginSet{},
-				Bind:       &PluginSet{},
-				PostBind:   &PluginSet{},
+				PostFilter: PluginSet{},
```
Comment (alculquicondor) [inline pkg/scheduler/apis/config/types_test.go:59]: so now you can actually remove these lines that have empty PluginSet :)

File: pkg/scheduler/apis/config/types_test.go (hunk 3 above)
Comment (alculquicondor) [reply pkg/scheduler/apis/config/types_test.go:243]: Uhmmm could it be that `append(nil, (nil)...)` in `appendPluginSet` is creating a pointer? You could experiment in https://play.golang.org/

Comment (arghya88) [reply pkg/scheduler/apis/config/types_test.go]: I have done some changes..PTAL..The test now passes.

File: pkg/scheduler/apis/config/types_test.go (hunk 6 above)
Comment (arghya88) [reply pkg/scheduler/apis/config/types_test.go]: Removed

**Comment by arghya88** [comment]
@alculquicondor So what we do for the type conversion? The tests are probably failing because of type conversion.

**Comment by alculquicondor** [comment]
`make generated_files`

**Comment by arghya88** [comment]
@alculquicondor When I do that `pkg/scheduler/apis/config/v1beta1/zz_generated.conversion.go` gets updated with 

```
// WARNING: in.QueueSort requires manual conversion: inconvertible types (*k8s.io/kube-scheduler/config/v1beta1.PluginSet vs k8s.io/kubernetes/pkg/scheduler/apis/config.PluginSet)
...
```

**Comment by alculquicondor** [comment]
I would have expected the automatic conversion to be able to handle pointer to non-pointer conversion. Maybe we can get rid of the pointers in staging too. I think this should be backwards compatible still.

Could you give it a try? We can determine if it's safe to do the change from the tests that will need updates.

**Comment by alculquicondor** [comment]
Actually, it's not safe to do the change, as anybody who is using the (staging) go types directly would need to change their code.

You have to provide a manual conversion function in conversion.go, similar to https://github.com/kubernetes/kubernetes/blob/2046f4212a355a3381da20c4d800283b98e9a081/pkg/scheduler/apis/config/v1beta1/conversion.go#L49

but for PluginSet and its pointer. See the `zz_generated.conversion.go` to figure what should be the name of the function.

**Comment by alculquicondor** [comment]
code freeze is approaching. Any progress on this?

**Comment by arghya88** [comment]
@alculquicondor working on it..Is there any example of pointer to non pointer conversion for struct type as opposed to above examples of primitive types?

**Comment by alculquicondor** [comment]
I don't think so, only for primitive types indeed https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/conversion.go

**Comment by alculquicondor** [comment]
Any progress on this? Code freeze is approaching

**Comment by arghya88** [comment]
> Any progress on this? Code freeze is approaching

No.I could not figure out the conversion.

**Comment by alculquicondor** [comment]
You can just provide the conversion function yourself. There are only 2 fields to copy over.

**Comment by arghya88** [comment]
@alculquicondor the conversion functions signature looks like below with both in and out being pointers. Is this correct?

```
func Convert_v1beta1_Plugins_To_config_Plugins(in *v1beta1.Plugins, out *config.Plugins, s conversion.Scope) error {

}

func Convert_config_Plugins_To_v1beta1_Plugins(in *config.Plugins, out *v1beta1.Plugins, s conversion.Scope) error {

}
```

**Comment by alculquicondor** [comment]
It might be. It's a matter of trying.
